### PR TITLE
Fix display of images created by the jupyter engine

### LINKED ISSUE #9470: Retina images generated using ipython display functions do not have their dimensions set
### Bug description

Images generated via jupyter/ipython and tagged as `retina` should be displayed in half the true pixel dimensions. For html format, the image tag should contain the `height` and `width` that are half the true size.

Generally, it is not possible to control the display size (and therefore the perceived resolution) of the output image using ipython display functions.

### Steps to reproduce

````qmd
---
title: "IPython Retina Image"
format: html
---

```{python}
from IPython.display import Image, display_png

url = "https://quarto.org/docs/gallery/websites/websites-quarto.png"
img = Image(url=url, metadata={"height": 663, "width": 900}, retina=True, embed=True)
display_png(img)
```
````

### Expected behavior

A a figure with image tag that has `height` and `width`. i.e.

```html
<img src="..." height="331" width="450" ...>
```

### Actual behavior

```html
<img src="..." class="img-fluid figure-img">
```

### Quarto check output

```bash
$ quarto check
Quarto 1.5.31
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.13: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.31
      Path: /Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: v2022.11
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /Users/me/Library/TinyTeX/bin/universal-darwin
      Version: 2022

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.12.1
      Path: /Users/me/.uvenv/plotnine/bin/python3
      Jupyter: 5.7.2
      Kernels: gnuplot, python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.2.3
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.2/Resources/library
      knitr: 1.45
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK
```

### PR DESCRIPTION
**Pull request by has2k1**
This PR makes quarto render Jupyter generated images to the same size as they are rendered in Jupyter native environments like jupyter-lab or notebook.

1. https://github.com/quarto-dev/quarto-cli/pull/9538/commits/25c725bd06cb27b2a61275b5961fd206c4c517c5 fixes an edge case for matplotlib images where the DPI of a displayed figure may not match the DPI with which the image is saved; contrary to the intention of the implementation.
2. https://github.com/quarto-dev/quarto-cli/pull/9538/commits/f835a7671963f2ebaa2ab024b5ad764d7ba7fe74 makes quarto consider the display width & height of generated images (if jupyter/ipython has set them) before falling back to the "fluid" image size.

References: https://github.com/has2k1/plotnine/issues/773, #9470

### CODE REVIEW

**Comment by cscheid** [comment]
We also need tests that check for the output so we don't regress later. You can pattern-match on the examples we have in `tests/docs/smoke-all`. You'll want to add some entries like

````yml
_quarto:
  tests:
    html:
      ensureFileRegexMatches:
        ...
      ensureHtmlElements:
        ...
````

Use `ensureFileRegexMatches` if it's easier to check with a regex, but more prefereably, use `ensureHtmlElements` which will check using `document.querySelector`.

**Comment by cscheid** [comment]
Tests pass 🎉 - let me know if you need help with the `smoke-all` stuff or the .ts fixes ahead of merging.

**Comment by cscheid** [comment]
(@has2k1 I added a merge commit for the changelog so that we can start running the test suite)

Those tests look good, thank you!

File: src/core/jupyter/jupyter.ts [hunk 1]
```diff
@@ -1909,11 +1909,14 @@ function mdImageOutput(
...
+  function metadataValue<T>(key: string, alt_key: string, defaultValue: T) {
+    if (metadata) {
+      return metadata[key] ? metadata[key] as T : metadata[alt_key] ? metadata[alt_key] as T : defaultValue;
+    }
+    return defaultValue;
   }
-  let width = metadataValue(kCellOutWidth, 0);
-  let height = metadataValue(kCellOutHeight, 0);
+  let width = metadataValue(kCellOutWidth, "width", 0);
+  let height = metadataValue(kCellOutHeight, "height", 0);
```
Comment (has2k1) [inline src/core/jupyter/jupyter.ts:1919]: I couldn't think of a case were the `metadata` would contain `KCellOutWidth="out-width"` & `KCellOutHeight="out-height"`. Is this for future support?

Comment (cderv) [reply src/core/jupyter/jupyter.ts:1919]: Sorry I missed this. I don't know exactly - I'll look closer. Maybe it was placeholder in anticipation to bring `out-width` and `out-height` to Jupyter (as it is supported in **knitr** initially)

**Pull request merged**